src/createQueryClient.spec-d.ts

```diff
@@ -22,4 +22,108 @@ describe('options', () => {
     query(action, [], { tags: [untypedTag] })
     query(action, [], { tags: () => [untypedTag] })
   })
+})
+
+describe('setQueryData', () => {
+  test('tags', async () => {
+    const { setQueryData } = createQueryClient()
+    const numberTag = tag<number>()
+    const stringTag = tag<string>()
+    const untypedTag = tag()
+
+    setQueryData(untypedTag, (data) => {
+      expectTypeOf(data).toMatchTypeOf<unknown>()
+      return 'foo'
+    })
+
+    setQueryData(numberTag, (data) => {
+      expectTypeOf(data).toMatchTypeOf<number>()
+      return 2
+    })
+
+    setQueryData(stringTag, (data) => {
+      expectTypeOf(data).toMatchTypeOf<string>()
+      return 'new string'
+    })
+
+    setQueryData([untypedTag], (data) => {
+      expectTypeOf(data).toMatchTypeOf<unknown>()
+      return 'foo'
+    })
+
+    setQueryData([numberTag], (data) => {
+      expectTypeOf(data).toMatchTypeOf<number>()
+      return 2
+    })
+
+    // this is kinda interesting, no matter the data the return type is the union :thinking:
+    // so there's not really a type safe way to update multiple queries at once
+    setQueryData([numberTag, stringTag], (data) => {
+      expectTypeOf(data).toMatchTypeOf<number | string>()
+      return 'foo'
+    })
```

Comment:
This made me wonder if the tag types are still not quite right. Idk how to fix this, but the callback doesn't enforce the return type per query. It just enforces that the return matches one of the tags, which isn't necessary correct on a per query basis. Might be able to improve that somehow, but drawing a blank.